doc/zeitgeist-daemon.1
.TH ZEITGEIST\-DAEMON 1 "June 28, 2012" "Zeitgeist"

.SH NAME
zeitgeist\-daemon \- D\-Bus interface providing a log of activities

.SH SYNOPSIS
\fBzeitgeist\-daemon\fP \fI[OPTION]\fP

.SH DESCRIPTION
\fBzeitgeist\-daemon\fP is daemon providing an activity log. That is,
it consists of a central database for events such as file usage, browser
history, calendar entries, etc.
.PP
All this information is made available to external applications which
communicate with \fBzeitgeist\-daemon\fP over D-Bus. Interested
applications can request information, insert new events or delete
existing events. 
.PP
Zeitgeist is also capable of establishing relationships between
activities. Its purpose is enabling an activity-aware desktop with
adaptive user interfaces.
.PP
By default, \fBzeitgeist\-daemon\fP will also start the bundled
logging application \fBzeitgeist\-datahub\fP.

.SH OPTIONS
The program follows the usual GNU command line syntax, with
options starting with two dashes (`--'). A summary of options is
included below.
.TP
.B \-\-no\-datahub
Do not start \fBzeitgeist\-datahub\fP together with this daemon.
.TP
.B \-r, \-\-replace
Replace any existing Zeitgeist instance if there is one already
running. If this option isn't used, \fBzeitgeist\-daemon\fP will
refuse to start if another running instance is found.
.TP
.B --log-level=\fILOG_LEVEL\fP
Specifies how much information should be printed to the standard output.
Possible values are: DEBUG, INFO, WARNING, ERROR, CRITICAL.
.TP
.B --log-file=\fILOG_FILE\fP
Specifies a file to which log output will be appended.
.TP
.B \-q, \-\-quit
Stop a currently running Zeitgeist instance and exit. This lets
Zeitgeist finish any operation it may currently be doing.
.TP
.B \-v, \-\-version
Shows which version of Zeitgeist is being executed.
.TP
.B \-h, \-\-help
Shows a short help message listing all supported options.

.SH ENVIRONMENT VARIABLES
.TP
.B ZEITGEIST_DATA_PATH
Overrides the directory where user data files are placed. The default
value is that of XDG's data dir (usually ~/.local/share/zeitgeist).
.TP
.B ZEITGEIST_DATABASE_PATH
This variable lets you specify an override for the filename of the main
database. The default value is $ZEITGEIST_DATA_PATH/activity.sqlite.
.TP
.B ZEITGEIST_DATABASE_BACKUP_PATH
This variable lets you specify an override for the filename of a possible
database backup. The default value is $ZEITGEIST_DATA_PATH/activity.sqlite.bck.
.TP
.B ZEITGEIST_DISABLED_EXTENSIONS
This variable lets you specify comma separated list of extension names which
shouldn't be loaded. The names are extracted from type names of the extensions,
so to disable extension with type name ZeitgeistDataSourceRegistry, value
of the variable should be DataSourceRegistry.

.SH SIGNALS
\fISIGHUP\f1: zeitgeist\-daemon will shut itself down in a clean way. 
.RS
It is recommended that system updates of Zeitgeist send this signal to 
all running instances of zeitgeist-daemon. DBus activation will make sure 
the updated software is automatically restarted when needed.

.SH EXIT CODES
.TP
.B 0
Zeitgeist terminated normally.
.TP
.B 1
An unspecified error occurred.
.TP
.B 10
There is already a running Zeitgeist instance.
.TP
.B 21
Could not access the database file.
.TP
.B 22
The database file is locked.

.SH SEE ALSO
\fBzeitgeist-datahub\fR, \fBgnome-activity-journal\fR

.SH BUGS
Please report any bugs on 
https://gitlab.freedesktop.org/zeitgeist/zeitgeist/issues.

.SH AUTHORS
Please see the AUTHORS file bundled with this application for
a complete list of contributors.

.SH LICENSE
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 2.1 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.
.PP
You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


NEWS
2020-10-15: Zeitgeist 1.0.3 "Ruby"
---------------------------------------------------------------

Overall:
 - Fix compile error with new vala
 - Port to python3 and update tests
 - Drop support for python2


2019-01-31: Zeitgeist 1.0.2 "Phoenix"
---------------------------------------------------------------

Overall:
 - Fix compile errors and warnings with new vala
 - fts: Correctly implement the org.gnome.zeitgeist.Index DBus interface
 - test-log: Use correct free function for ZeitgeistResultSet
 - Do not require rapper as it is not used anymore
 - Make ontology2code work with python3
 - Properly fallback with the Zeitgeist version
 - rdflib directly to parse the trig files
 - Update Ontology2code to ba able to specify the output path

Libzeitgeist:
 - Asynchronous out-parameters are only allowed at the end of argument lists
   *** THIS IS AN UNAVOIDABLE VALA API BREAK ***


2018-02-14: Zeitgeist 1.0.1 "Sunset"
---------------------------------------------------------------

Overall:
 - Fix placeholder in systemd service template

Libzeitgeist:
 - Fix some vala compile warnings


2017-02-18: Zeitgeist 1.0 "Epiphany"
---------------------------------------------------------------

Overall:
 - Drop dependency on gnome-common and intltool
 - Use GNU gettext only
 - Fix some valac warnings
 - Use $PYTHON to check for rdflib
 - Install zeitgeist-fts to pkglibexecdir rather than
   libexecdir
 - Use correct names for D-Bus service files
 - Add systemd user service for each D-Bus session service
 - Fix bash auto-completions script
 - Add --vacuum option to Daemon to perform VACUUM SQLite
   database

Libzeitgeist:
 - Ensure the log mainloop uses the current thread's context
   when waiting


2015-07-08: Zeitgeist 0.9.16 (1.0 Beta 8) "Experience"
---------------------------------------------------------------

Overall:
 - Don't hardcode python path
 - Fix build with vala-0.24+
 - Require glib 2.35.4 and adjust target-glib accordingly

Libzeitgeist:
 - Add -fPIC to libzeitgeist-internal.la and
   libzeitgeist_datamodel-2.0.la
 - Log: Need to quit the loop also on error, otherwise we
   leave a stray main loop around.


2014-01-15: Zeitgeist 0.9.15 (1.0 Beta 7) "Bazooka"
---------------------------------------------------------------

Libzeitgeist:
 - Fix converting variants to event lists

Datahub:
 - change some warnings to debugs


2013-06-30: Zeitgeist 0.9.14 (1.0 Beta 6) "Diamond"
---------------------------------------------------------------

Libzeitgeist:
 - Fix managing and tracking registered Monitors


2013-05-29: Zeitgeist 0.9.13 (1.0 Beta 5) "Silence is Gold"
---------------------------------------------------------------

Engine:
 - Adjust resetting of statement to proper sequence
 - Add custom memory PRAGMA to improve memory consumption

Libzeitgeist:
 - Add default memory PRAGMA to be used by clients

FTS:
 - Improve FTS memory consumption

Overall:
 - Fix compilation breaks for --enable-datahub --disable-downloads-monitor

This release is the fifth beta which will lead to zeitgeist 1.0

2013-04-12: Zeitgeist 0.9.12 (1.0 Beta 4) "Happy Go Lucky"
---------------------------------------------------------------

Libzeitgeist:
 - Fix find_related_uris to only print out the ranking if
   EXPLAIN_QUERIES is enabled
 - Fixed find_related_uris to take RelevantResultType as an
   argument instead of ResultType
 - Pass copy of templates to search and search_with_relevancies

Overall:
 - Fix documentation

This release is the fourth beta which will lead to zeitgeist 1.0

2013-04-09: Zeitgeist 0.9.11 (1.0 Beta 3) "H2O"
---------------------------------------------------------------

Libzeitgeist:
 - Fix the log's insert_events_no_reply and insert_event_no_reply
   API to not require an async callback
 - Pass copy of templates to threads in get_events, find_events,
   find_event_ids and find_related_uris
 - Pass copy of templates to search and search_with_relevancies
 - Expose RemoteLog's extensions and datapath properties via Log
   class

Overall:
 - Fix documentation title to Zeitgeist Reference Manual

This release is the third beta which will lead to zeitgeist 1.0

2013-03-19: Zeitgeist 0.9.10 (1.0 Beta 2) "Isis"
---------------------------------------------------------------

This version includes zeitgeist datahub in the release source 
instead of a separate project which was called zeitgeist-datahub

You can explicitly disable datahub by passing --enable-datahub=no 
and if you just want the telepathy datasource to be disabled you 
can just pass --disable-telepathy along with --enable-datahub

For datahub, json-glib 0.14.0 or greater is needed and for 
telepathy telepathy-glib 0.18.0 or greater is needed

Libzeitgeist:
 - Add new zg#ApplicationAction subject interpretation
 - Fix sql explanation bug

This release is the second beta which will lead to zeitgeist 1.0

2013-03-12: Zeitgeist 0.9.9 (1.0 Beta 1) "Pinch Me"
---------------------------------------------------

This version introduces libzeitgeist2, a Vala port of the previously
independent libzeitgeist library. Now everything you need to work with
Zeitgeist, be it with C, Vala or Python, is included in this tarball.

Please note that starting from this release Vala 0.18 is required to
build Zeitgeist.

Engine:
 - Fixed insertion of large event sets.
 - Fixed database migration to WebDataObject (from some RemoteDataObjects).
 - Events missing required meta-data are now rejected.

FTS:
 - Fixed to respect ZEITGEIST_DATA_PATH.

Python API:
 - RelevantResultType is now included in "from datamodel import *".

If you were using the external libzeitgeist, you need to be aware that
libzeitgeist2 has some API differences:
 - For Vala developers: getters/setters replaced with properties, etc.
 - For C developers: the use of floating references has been removed.
 - Redesigned ResultSet (removed peek() and seek(); introduced reset();
   renamed next() to next_value()).

zeitgeist-daemon.py
#! /usr/bin/env python
# -.- coding: utf-8 -.-

# Zeitgeist
#
# Copyright © 2009 Siegfried-Angel Gevatter Pujals <rainct@ubuntu.com>
# Copyright © 2010 Markus Korn <thekorn@gmx.de>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 2.1 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import sys
import os
import gobject
import glib
import dbus.mainloop.glib
import gettext
import logging
import logging.handlers
import optparse
import signal
from subprocess import Popen, PIPE
from xdg import BaseDirectory
from time import time

# Make sure we can find the private _zeitgeist namespace
from zeitgeist import _config
_config.setup_path()

# Make sure we can load user extensions, and that they take priority over
# system level extensions
from _zeitgeist.engine import constants
sys.path.insert(0, constants.USER_EXTENSION_PATH)

gettext.install("zeitgeist", _config.localedir, unicode=True)

class Options(optparse.Option):
	TYPES = optparse.Option.TYPES + ("log_levels",)
	TYPE_CHECKER = optparse.Option.TYPE_CHECKER.copy()
	log_levels = ("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL")

	def check_loglevel(option, opt, value):
		value = value.upper()
		if value in Options.log_levels:
			return value
		raise optparse.OptionValueError(
			"option %s: invalid value: %s" % (opt, value))
	TYPE_CHECKER["log_levels"] = check_loglevel


def which(executable):
	""" helper to get the complete path to an executable """
	p = Popen(["which", str(executable)], stderr=PIPE, stdout=PIPE)
	p.wait()
	return p.stdout.read().strip() or None

def parse_commandline():
	parser = optparse.OptionParser(version = _config.VERSION, option_class=Options)
	parser.add_option(
		"-r", "--replace",
		action="store_true", default=False, dest="replace",
		help=_("if another Zeitgeist instance is already running, replace it"))
	parser.add_option(
		"--no-datahub", "--no-passive-loggers",
		action="store_false", default=True, dest="start_datahub",
		help=_("do not start zeitgeist-datahub automatically"))
	parser.add_option(
		"--log-level",
		action="store", type="log_levels", default="DEBUG", dest="log_level",
		help=_("how much information should be printed; possible values:") + \
			" %s" % ", ".join(Options.log_levels))
	parser.add_option(
		"--quit",
		action="store_true", default=False, dest="quit",
		help=_("if another Zeitgeist instance is already running, replace it"))
	parser.add_option(
		"--shell-completion",
		action="store_true", default=False, dest="shell_completion",
		help=optparse.SUPPRESS_HELP)
	return parser

def do_shell_completion(parser):
	options = set()
	for option in (str(option) for option in parser.option_list):
		options.update(option.split("/"))
	print " ".join(options)
	return 0

def setup_interface():
	from _zeitgeist.engine.remote import RemoteInterface
	dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
	mainloop = gobject.MainLoop()
	return mainloop, RemoteInterface(mainloop = mainloop)

def start_datahub():
	DATAHUB = "zeitgeist-datahub"
	# hide all output of the datahub for now,
	# in the future we might want to be more verbose here to make
	# debugging easier in case sth. goes wrong with the datahub
	devnull = open(os.devnull, "w")
	try:
		# we assume to find the datahub somewhere in PATH
		pid, stdin, stdout, stderr = glib.spawn_async(
			[DATAHUB,],
			flags=glib.SPAWN_SEARCH_PATH | glib.SPAWN_STDOUT_TO_DEV_NULL | glib.SPAWN_STDERR_TO_DEV_NULL)
	except (OSError, glib.GError):
		logging.warning("Unable to start the datahub, no binary found")
	else:
		# TODO: delayed check if the datahub is still running after some time
		#  and not failed because of some error
		# tell the user which datahub we are running
		logging.debug("Running datahub (%s) with PID=%i" %(which(DATAHUB), pid))

def setup_handle_exit(interface):
	def handle_exit(signum=None, frame=None):
		""" We want to end Zeitgeist in a clean way when we receive a
		SIGTERM or SIGHUP signal, or the user pressed Ctrl+C. """
		logging.info("Shutting down Zeitgeist interface...")
		interface.Quit()
	return handle_exit

class CustomFileLogFormatter(logging.Formatter):

	def __init__(self):
		format = "[%(asctime)s] - %(levelname)s - %(name)s - %(message)s"
		logging.Formatter.__init__(self, format)
		self._pid = os.getpid()

	def format(self, record):
		return "%s %s" % (self._pid, logging.Formatter.format(self, record))

def setup_logger(log_level):
	logger = logging.getLogger()
	logger.setLevel(getattr(logging, log_level))
	
	stream_handler = logging.StreamHandler()
	stream_handler.setFormatter(logging.Formatter(
		"[%(levelname)s - %(name)s] %(message)s"))
	logger.addHandler(stream_handler)
	
	try:
		log_file = os.environ["ZEITGEIST_LOG_FILE"]
	except KeyError:
		log_file = constants.DEFAULT_LOG_PATH
		if not os.path.exists(os.path.dirname(log_file)):
			os.makedirs(os.path.dirname(log_file))
	try:
		file_handler = logging.handlers.TimedRotatingFileHandler(log_file,
			when="midnight", backupCount=3)
		file_handler.setFormatter(CustomFileLogFormatter())
		logger.addHandler(file_handler)
	except IOError, e:
		logging.warning("Can't log to %s: %s" % (e.filename, e.strerror))

def main():
	
	parser = parse_commandline()
	
	_config.options, _config.arguments = parser.parse_args()
	if _config.options.shell_completion:
		sys.exit(do_shell_completion(parser))
	
	setup_logger(_config.options.log_level)
	
	try:
		mainloop, interface = setup_interface()
	except RuntimeError, e:
		logging.critical("Failed to setup the RemoteInterface")
		logging.info(e.args[0])
		sys.exit(1)
	
	if _config.options.start_datahub:
		logging.info("Trying to start the datahub")
		start_datahub()
	
	handle_exit = setup_handle_exit(interface)
	
	signal.signal(signal.SIGHUP, handle_exit)
	signal.signal(signal.SIGTERM, handle_exit)
	
	logging.info("Starting Zeitgeist service...")
	try:
		mainloop.run()
	except KeyboardInterrupt:
		handle_exit()

if __name__ == "__main__":
	main()
